ext/objspace/objspace_dump.c

```diff
@@ -19,38 +19,158 @@
+            // TODO: raise?
+        }
+    }
+}
+
+static void buffer_append(struct dump_config *dc, const char *cstr, unsigned long len)
+{
+    if (len > 0) {
+        buffer_ensure_capa(dc, len);
+        MEMCPY(dc->buffer + dc->buffer_len, cstr, char, len);
+        dc->buffer_len += len;
+    }
+}
+
+# define dump_append(dc, str) buffer_append(dc, (str), (long)strlen(str))
+
+static void
+dump_append_ld(struct dump_config *dc, const long number)
+{
+    const int width = DECIMAL_SIZE_OF_BITS(sizeof(number) * CHAR_BIT - 1) + 1;
+    buffer_ensure_capa(dc, width);
+    unsigned long required = snprintf(dc->buffer + dc->buffer_len, width, "%ld", number);
+    RUBY_ASSERT(required <= width);
+    dc->buffer_len += required;
+}
+
+static void
+dump_append_lu(struct dump_config *dc, const unsigned long number)
+{
+    const int width = DECIMAL_SIZE_OF_BITS(sizeof(number) * CHAR_BIT);
+    buffer_ensure_capa(dc, width);
+    unsigned long required = snprintf(dc->buffer + dc->buffer_len, width, "%lu", number);
+    RUBY_ASSERT(required <= width);
+    dc->buffer_len += required;
+}
+
+static void
+dump_append_g(struct dump_config *dc, const double number)
 {
-    va_list vl;
-    va_start(vl, format);
+    unsigned long capa_left = BUFFER_CAPACITY - dc->buffer_len;
+    unsigned long required = snprintf(dc->buffer + dc->buffer_len, capa_left, "%#g", number);
 
-    if (dc->stream) {
-	vfprintf(dc->stream, format, vl);
+    if (required >= capa_left) {
+        buffer_ensure_capa(dc, required);
+        capa_left = BUFFER_CAPACITY - dc->buffer_len;
+        snprintf(dc->buffer + dc->buffer_len, capa_left, "%#g", number);
     }
-    else if (dc->string)
-	rb_str_vcatf(dc->string, format, vl);
+    dc->buffer_len += required;
+}
+
+static void
+dump_append_d(struct dump_config *dc, const int number)
+{
+    const int width = DECIMAL_SIZE_OF_BITS(sizeof(number) * CHAR_BIT - 1) + 1;
+    buffer_ensure_capa(dc, width);
+    unsigned long required = snprintf(dc->buffer + dc->buffer_len, width, "%d", number);
+    RUBY_ASSERT(required <= width);
+    dc->buffer_len += required;
+}
 
-    va_end(vl);
+static void
+dump_append_sizet(struct dump_config *dc, const size_t number)
+{
+    const int width = DECIMAL_SIZE_OF_BITS(sizeof(number) * CHAR_BIT);
+    buffer_ensure_capa(dc, width);
+    unsigned long required = snprintf(dc->buffer + dc->buffer_len, width, "%"PRIuSIZE, number);
+    RUBY_ASSERT(required <= width);
+    dc->buffer_len += required;
+}
+
+static void
+dump_append_c(struct dump_config *dc, char c)
+{
+    if (c <= 0x1f) {
+        buffer_ensure_capa(dc, 6);
+        sprintf(dc->buffer + dc->buffer_len, "\\u%04x", c);
+    }
+    else {
+        buffer_ensure_capa(dc, 1);
+        dc->buffer[dc->buffer_len] = c;
+        dc->buffer_len++;
+    }
+}
+
+static const char *hex_table = "0123456789abcdef";
+
+static void
+dump_append_ref(struct dump_config *dc, VALUE ref)
+{
+    char buffer[20];
```

Comment:
Should not assume platform dependent value.
```suggestion
    char buffer[(sizeof(VALUE) * CHAR_BIT + 3)/ 4];
```

Reply:
Thanks, applied.